### ISSUE TITLE
Ruby 3.2.1 SideKiq error - LinkCrawlWorker

### ISSUE THREAD

**Issue by supernovae** [mastodon/mastodon#23637]
### Steps to reproduce the problem

Upgraded to Ruby 3.2.1 and everything started up, but I'm seeing errors in Recurring jobs

```
NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
```

![image](https://user-images.githubusercontent.com/1461719/219434775-b39496ab-c1fa-450f-a941-9d6606830c2a.png)

### Expected behaviour

Jobs Complete

### Actual behaviour

```
 NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
```

### Detailed description

sidekiq LinkCrawlWorker

### Specifications

Mastodon 4.1.0 from main branch
Ruby 3.2.1 

```
sudo apt install rustc

su - mastodon

cd ~/.rbenv 
git pull
cd plugins/ruby-build
git pull

cd ~/.rbenv && src/configure && make -C src

RUBY_CONFIGURE_OPTS=--enable-yjit rbenv install 3.2.1

rbenv global 3.2.1
rbenv local 3.2.1

exit
su - mastodon

cd ~/live

git fetch upstream main  # (you will need to pull down main mastodon branch, i call it upstream)

bundle install && yarn install

RAILS_ENV=production bundle exec rails assets:precompile

systemctl restart mastodon-web # (on web services)
systemctl restart mastodon-sidekiq* # (for sidekiq instances)
```

**Comment by tisba** [comment]
Don't want to create unnecessary noise as I'm not sure (sorry in advance), but this could be related to https://github.com/mastodon/mastodon/pull/22928#issuecomment-1431706224. There is a GC related issue with bootsnap in Ruby 3.1 and 3.2 that can cause weird errors like this one. There is a fix which needs to be backported, see https://bugs.ruby-lang.org/issues/19419.

**Comment by tenderlove** [comment]
Is there a way to get a full stack trace?  I will try to reproduce locally, but I don't have any real data to work with so I'm kind of guessing.

**Comment by tenderlove** [comment]
@tisba I don't think that bug is applicable.  It only occurs when dumping iseqs as binary, which it looks like [Mastodon does not do](https://github.com/tenderlove/mastodon/blob/68a92c81bfb1bc4703cd4974321b8cb168793541/config/boot.rb#L15).

My guess is something to do with marshal data format between Ruby versions, but it is a very vague guess given we don't have a full stack trace and I can't reproduce it locally (yet).

**Comment by supernovae** [comment]
I'll turn on debug logging and try and get a full stack trace

**Comment by supernovae** [comment]
Is this enough?  not very big output/trace

eb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.623Z pid=3508908 tid=254k8 WARN: {"context":"Job raised exception","job":{"retry":0,"queue":"pull","class":"LinkCrawlWorker","args":[109875246184843283],"jid":"63e31bb5617e0ef419cdddcb","created_at":1676563205.2537515,"enqueued_at":1676577100.6246097,"error_message":"undefined method `_read_attribute' for :read_attribute_for_validation:Symbol","error_class":"NoMethodError","failed_at":1676563264.1300058,"retry_count":-1}}
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.623Z pid=3508908 tid=254k8 WARN: NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.624Z pid=3508908 tid=254k8 WARN: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/attribute_methods.rb:254:in `image_content_type'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/bundler/gems/kt-paperclip-11abf222dc31/lib/paperclip/validators/attachment_content_type_validator.rb:19:in `validate_each'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/validator.rb:153:in `block in validate'
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.620Z pid=3508908 tid=24bc4 class=LinkCrawlWorker jid=dbf874f4c8ce94180a1b1e9b elapsed=1.033 INFO: fail
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.624Z pid=3508908 tid=24bc4 WARN: {"context":"Job raised exception","job":{"retry":0,"queue":"pull","class":"LinkCrawlWorker","args":[109876151266138383],"jid":"dbf874f4c8ce94180a1b1e9b","created_at":1676577015.2407317,"enqueued_at":1676577100.8559854,"error_message":"undefined method `_read_attribute' for :read_attribute_for_validation:Symbol","error_class":"NoMethodError","failed_at":1676577064.9042256,"retry_count":-1,"retried_at":1676577093.675221}}
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.624Z pid=3508908 tid=24bc4 WARN: NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.624Z pid=3508908 tid=24bc4 WARN: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/attribute_methods.rb:254:in `image_content_type'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/bundler/gems/kt-paperclip-11abf222dc31/lib/paperclip/validators/attachment_content_type_validator.rb:19:in `validate_each'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/validator.rb:153:in `block in validate'
Feb 16 19:54:11 amdworker bundle[3508908]:   PreviewCard Exists? (3.1ms)  SELECT 1 AS one FROM "preview_cards" INNER JOIN "preview_cards_statuses" ON "preview_cards"."id" = "preview_cards_statuses"."preview_card_id" WHERE "preview_cards_statuses"."status_id" = 109876070748656011 LIMIT 1
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.609Z pid=3508908 tid=24ji0 class=LinkCrawlWorker jid=97c1e2237f33b4cf6f365a0a INFO: start
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.608Z pid=3508908 tid=24bxw class=LinkCrawlWorker jid=7be4de80cdb0978bedd28947 INFO: Adding dead LinkCrawlWorker job 7be4de80cdb0978bedd28947
Feb 16 19:54:11 amdworker bundle[3508908]:   Status Load (1.5ms)  SELECT "statuses".* FROM "statuses" WHERE "statuses"."deleted_at" IS NULL AND "statuses"."id" = 109876071152869359 ORDER BY "statuses"."id" DESC LIMIT 1
Feb 16 19:54:11 amdworker bundle[3508908]:   PreviewCard Load (1.0ms)  SELECT "preview_cards".* FROM "preview_cards" WHERE "preview_cards"."url" = 'https://www.mirror.co.uk/sport/football/news/man-utd-takeover-saudi-arabia-29237112?t=' LIMIT 1
Feb 16 19:54:11 amdworker bundle[3508908]:   Status Load (2.1ms)  SELECT "statuses".* FROM "statuses" WHERE "statuses"."deleted_at" IS NULL AND "statuses"."id" = 109876071133828980 ORDER BY "statuses"."id" DESC LIMIT 1
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.625Z pid=3508908 tid=24jds class=LinkCrawlWorker jid=333d3b250b429de74c0e5da5 INFO: start
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.617Z pid=3508908 tid=24brg class=LinkCrawlWorker jid=5287703be9d1d04f6f1d585a elapsed=1.196 INFO: fail
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.629Z pid=3508908 tid=24brg WARN: {"context":"Job raised exception","job":{"retry":0,"queue":"pull","class":"LinkCrawlWorker","args":[109876155215851011],"jid":"5287703be9d1d04f6f1d585a","created_at":1676577075.468061,"enqueued_at":1676577100.8514674,"error_message":"undefined method `_read_attribute' for :read_attribute_for_validation:Symbol","error_class":"NoMethodError","failed_at":1676577089.1485271,"retry_count":-1,"retried_at":1676577093.7387989}}
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.629Z pid=3508908 tid=24brg WARN: NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.629Z pid=3508908 tid=24brg WARN: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/attribute_methods.rb:254:in `image_content_type'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/bundler/gems/kt-paperclip-11abf222dc31/lib/paperclip/validators/attachment_content_type_validator.rb:19:in `validate_each'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/validator.rb:153:in `block in validate'
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.626Z pid=3508908 tid=24jd8 class=LinkCrawlWorker jid=aa826e838c64104a9f8ac5df INFO: start
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.629Z pid=3508908 tid=24jg0 class=LinkCrawlWorker jid=9f9be783e8cdca4c767181de INFO: start
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.626Z pid=3508908 tid=24bxw class=LinkCrawlWorker jid=7be4de80cdb0978bedd28947 elapsed=1.124 INFO: fail
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.630Z pid=3508908 tid=24bxw WARN: {"context":"Job raised exception","job":{"retry":0,"queue":"pull","class":"LinkCrawlWorker","args":[109876152237367866],"jid":"7be4de80cdb0978bedd28947","created_at":1676577030.2189734,"enqueued_at":1676577100.8533182,"error_message":"undefined method `_read_attribute' for :read_attribute_for_validation:Symbol","error_class":"NoMethodError","failed_at":1676577074.2341259,"retry_count":-1,"retried_at":1676577093.7003949}}
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.630Z pid=3508908 tid=24bxw WARN: NoMethodError: undefined method `_read_attribute' for :read_attribute_for_validation:Symbol
Feb 16 19:54:11 amdworker bundle[3508908]: 2023-02-16T19:54:11.630Z pid=3508908 tid=24bxw WARN: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/attribute_methods.rb:254:in `image_content_type'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/bundler/gems/kt-paperclip-11abf222dc31/lib/paperclip/validators/attachment_content_type_validator.rb:19:in `validate_each'
Feb 16 19:54:11 amdworker bundle[3508908]: /home/mastodon/live/vendor/bundle/ruby/3.2.0/gems/activemodel-6.1.7.2/lib/active_model/validator.rb:153:in `block in validate'

**Comment by tenderlove** [comment]
@supernovae thanks, this is a big help.  Looks like something to do with kt-paperclip and the `PreviewCard` model.  I'd guess this is due to the `kt-paperclip` upgrade and not to do with the Ruby upgrade.

The `FetchLinkCard` service seems to creates new `PreviewCard` models.  The `PreviewCard` model has a `image_content_type`field (which lines up with the messages in the stack trace).  Looks like active model is trying to call `_read_attribute` on the `:read_attribute_for_validation` symbol which came from [here](https://github.com/kreeti/kt-paperclip/blob/11abf222dc31bff71160a1d138b445214f434b2b/lib/paperclip/validators/attachment_content_type_validator.rb#L19).

I don't see any tests for the `PreviewCard` model.  I'll try to figure out how to instantiate one and hopefully that will point us to the problem.

Thanks again for providing the output, it's very helpful!

**Comment by tenderlove** [comment]
I tried to reproduce this, but I can't seem to get anywhere.  I tried a script like this:

```ruby
email = 'aaron@example.org'

user = User.find_by_email email

unless user
  owner_role = UserRole.find_by(name: 'Owner')
  password = SecureRandom.hex(16)
  username = 'aaron'
  user = User.new(email: email, password: password, confirmed_at: Time.now.utc, account_attributes: { username: username }, bypass_invite_request_check: true, role: owner_role)
  user.save(validate: false)
end

text = SecureRandom.hex(16)

status = Status.new(account: user.account,
                    text: "#{text} https://github.com/mastodon/mastodon/issues/23637")
status.save!

service = FetchLinkCardService.new
service.call status

card = status.preview_cards.first
p card.send :read_attribute_for_validation, :image_content_type
```

I _think_ the `card` instance in my test script is equivalent to the `record` variable [found here](https://github.com/kreeti/kt-paperclip/blob/11abf222dc31bff71160a1d138b445214f434b2b/lib/paperclip/validators/attachment_content_type_validator.rb#L19).  But my test script is not raising an exception.

I'm unsure how to reproduce this locally.  I'm happy to jump on a call with you to try and debug this (if you have time in the next week or so).

**Comment by supernovae** [comment]
I think it's a me problem lol.  Had something capturing if something has a link that somehow broke - i'm researching now and this may be safe to close.  (i'll try and update by tomorrow - need to spend some time with fam hehe)

**Comment by ClearlyClaire** [comment]
I'm not able to reproduce the issue.

> Had something capturing if something has a link that somehow broke - i'm researching now and this may be safe to close.

I'm not sure what you mean: you had a custom code change somewhere in Mastodon? If so, can you share it so that we could see if it rules out other issues?

**Comment by supernovae** [comment]
Just as an update, i pulled the latest main and all is well. thank you so much for the insight @tenderlove - going to keep experimenting with moving 100% towards ruby 3.2.1/YJIT

**Comment by untitaker** [comment]
For what it's worth, this issue also definitely happened on my instance when I tried to enable yjit on `68a92c81bfb1bc4703cd4974321b8cb168793541`. It cannot however be reproduced in the testsuite. Curious whether you're able to roll out to 100% after pulling @supernovae

**Comment by supernovae** [comment]
I'll re-open

Mostly because I pulled a later main branch, and the sidekiq workers started working, but pulling main on webtier (puma) gave 500s for user uploads and I immediately reverted instead of further debugging.

**Comment by untitaker** [comment]
I think this should not be closed. I would love to see Mastodon working correctly with YJIT. Mastodon currently takes up too much CPU even for I/O bound tasks.

**Comment by untitaker** [comment]
oh @supernovae are you saying you are successfully running sidekiq with YJIT but are running into https://github.com/mastodon/mastodon/issues/23644?

**Comment by untitaker** [comment]
My path forward is to run a subset of sidekiq workers with YJIT to try and collect more data.

Running web workers with YJIT is highly inadvisable since I also have encountered this crash on `/inbox`.

**Comment by untitaker** [comment]
https://github.com/Shopify/yjit/issues/306

**Comment by Teutates** [comment]
Update: ruby 3.2.2 has been released, which includes the fix.
https://github.com/ruby/ruby/releases/tag/v3_2_2

**Comment by renchap** [comment]
Closing as this has been fixed in Ruby 3.2.2
